src/reorder.c

```diff
@@ -37,66 +37,61 @@ SEXP reorder(SEXP x, SEXP order)
 
   const int *restrict idx = INTEGER(order);
   int i=0;
-  while (i<nrow && idx[i] == i+1) i++;
+  while (i<nrow && idx[i] == i+1) ++i;
   const int start = i;
   if (start==nrow) { UNPROTECT(nprotect); return(R_NilValue); }  // input is 1:n, nothing to do
   i = nrow-1;
-  while (idx[i] == i+1) i--;
+  while (idx[i] == i+1) --i;
   const int end = i;
-  for (int i=start; i<=end; i++) {
+  for (int i=start; i<=end; ++i) {
     int itmp = idx[i]-1;
     if (itmp<start || itmp>end) error(_("order is not a permutation of 1:nrow[%d]"), nrow);
+    // This file is for internal use (so we should get the input right), but this check seems sensible to still perform since it should
+    // run in negligible time (sequential with prefetch). It is not sufficient though; e.g. a non-permutation such as a duplicate will not be caught.
+    // But checking for dups would incur too much cost given that i) this is for internal use only and we only use this function internally
+    // when we're sure we're passing in a strict permutation, and ii) this operation is fundamental to setkey and data.table.
   }
-  // Creorder is for internal use (so we should get the input right!), but the check above seems sensible. The for loop above should run
-  // in neglible time (sequential with prefetch). It will catch NAs anywhere but won't catch duplicates. But doing so would be going too
-  // far given this is for internal use only and we only use this function internally when we're sure it's a permutation.
 
-  char *TMP = malloc(nrow * maxSize);
-  // enough RAM for a copy of one column (of largest type). Writes into the [start,end] subset. Outside [start,end] is wasted in that rarer case
-  // to save a "-start" in the deep loop below in all cases.
+  char *TMP = malloc((end-start+1)*maxSize);
```

Comment:
Any reason why we don't use R_alloc here?

Reply:
Nope. Good spot.